### PR TITLE
SFT Pipeline for Maxtext-Omni

### PR DESCRIPTION
**Pull request by subawocit**
# Description

This PR introduces the native Supervised Fine-Tuning (SFT) training runner and checkpoint verification utilities for hybrid multimodal Omni models (Gemma 3 Vision + Qwen 3 LLM).

Specifically:
- **Two-Stage Multimodal Training Configurations**:
  1. **Stage 1 (Pretrain / Projector Alignment)**: `pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml` for pretraining the randomly initialized MLP connector on chart data (ChartNet).
  2. **Stage 2 (Downstream SFT)**: `sft-maxtext-omni-gemma3-qwen3.yml` for visual question-answering fine-tuning (ChartQA) starting from an aligned checkpoint.
  - Both configs are trained using `train_sft_omni.py`
- **Checkpoint Freezing Verification**: A test utility `compare_sft_checkpoint_test.py` that loads the base stitched checkpoint alongside the SFT-trained checkpoint, comparing all parameter PyTree leaves to mathematically verify that only projector weights changed while frozen encoder/decoder weights remained 100% identical.

This is Step 5 of a 5-step Proof-of-Concept for any-to-any multimodal alignment in MaxText. The overall goal is to align pretrained vision encoders with text-only LLMs by connecting the two with a dynamically configured adaptation layer, special tokenizer mapping for visual placeholder tokens, and training only the MLP using supervised fine tuning.

---

# Files

1. **Omni SFT Training Runner**
   - `src/maxtext/experimental/omni_poc/train_sft_omni.py`
     - Implements native SFT training loop execution for Omni models.

2. **Checkpoint Freezing & Weight Comparison Test**
   - `src/maxtext/experimental/omni_poc/tests/compare_sft_checkpoint_test.py`
     - Test to load and compare pre-SFT vs post-SFT checkpoints, validating that `Vision Encoder` and `LLM Backbone` parameters remain frozen while `Projector MLP` parameters are updated.

3. **Training Configurations**
   - `src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml`
     - Configuration for pretraining on ChartNet dataset.
   - `src/maxtext/experimental/omni_poc/configs/sft-maxtext-omni-gemma3-qwen3.yml`
     - Configuration for SFT on ChartQA dataset.

---

# Example Usage

## Two-Stage Training Example

```bash
# Stage 1: Pretrain Projector Alignment on ChartNet
python3 -m maxtext.experimental.omni_poc.train_sft_omni \
  src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml \
  load_parameters_path=gs://YOUR_BUCKET/path/to/stitched/checkpoint/items \
  base_output_directory=gs://YOUR_BUCKET/pretrain_output \
  run_name=my_chartnet_pretrain 

# Stage 2: SFT Fine-Tuning on ChartQA
python3 -m maxtext.experimental.omni_poc.train_sft_omni \
  src/maxtext/experimental/omni_poc/configs/sft-maxtext-omni-gemma3-qwen3.yml \
  load_parameters_path=gs://YOUR_BUCKET/pretrain_output/my_chartnet_pretrain/checkpoints/100/items \
  base_output_directory=gs://YOUR_BUCKET/sft_output \
  run_name=my_chartqa_sft 
```

---

# Training Execution & Verification

## 1. Stage 1: Pretraining on ChartNet
```bash
export PYTHONPATH=src:$PYTHONPATH

python3 -m maxtext.experimental.omni_poc.train_sft_omni \
  src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml \
  load_parameters_path="gs://yuchenhou-maxtext-logs/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items" \
  base_output_directory="gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/pretrain_chartnet" \
  run_name="omni_pretrain_chartnet_100step" \
  steps=100 \
  checkpoint_period=100 \
  eval_interval=0
```

## 2. Stage 2: SFT on ChartQA
```bash
export PYTHONPATH=src:$PYTHONPATH

python3 -m maxtext.experimental.omni_poc.train_sft_omni \
  src/maxtext/experimental/omni_poc/configs/sft-maxtext-omni-gemma3-qwen3.yml \
  load_parameters_path="gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/pretrain_chartnet/omni_pretrain_chartnet_100step/checkpoints/99/items" \
  base_output_directory="gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/sft_chartqa" \
  run_name="omni_sft_chartqa_100step" \
  steps=100 \
  checkpoint_period=100 \
  save_checkpoint_on_completion=true \
  eval_interval=0 \
  eval_steps=0
```

### Training Output:
```text
I0825 23:09:57.545643 139925043539072 metric_logger.py:230] completed step: 0, seconds: 77.607, TFLOP/s/device: 2.739, Tokens/s/device: 105.558, total_weights: 6413, loss: 1.372, lm_loss: 1.372, perplexity: 3.942
I0825 23:10:00.088244 139925043539072 metric_logger.py:230] completed step: 1, seconds: 7.287, TFLOP/s/device: 29.169, Tokens/s/device: 1124.246, total_weights: 6845, loss: 1.435, lm_loss: 1.435, perplexity: 4.198
I0825 23:10:02.636908 139925043539072 metric_logger.py:230] completed step: 2, seconds: 1.408, TFLOP/s/device: 150.968, Tokens/s/device: 5818.645, total_weights: 6478, loss: 1.344, lm_loss: 1.344, perplexity: 3.834
I0825 23:10:05.185393 139925043539072 metric_logger.py:230] completed step: 3, seconds: 1.380, TFLOP/s/device: 154.072, Tokens/s/device: 5938.289, total_weights: 7299, loss: 1.274, lm_loss: 1.274, perplexity: 3.577
```

---

# Tests

## Checkpoint Freezing Verification Test
Verify that only the custom MLP projector was trained and all encoder/decoder weights remained strictly frozen:
```bash
export PYTHONPATH=src:$PYTHONPATH

python3 -m maxtext.experimental.omni_poc.tests.compare_sft_checkpoint_test \
  src/maxtext/experimental/omni_poc/configs/sft-maxtext-omni-gemma3-qwen3.yml \
  load_parameters_path="gs://yuchenhou-maxtext-logs/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items" \
  base_output_directory="gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/sft_chartqa" \
  run_name="omni_sft_chartqa_100step"
```

### Verification Output:
```text
================================================================================
  Stitched Original Checkpoint: gs://yuchenhou-maxtext-logs/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items
  SFT Trained Checkpoint:       gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/sft_chartqa/omni_sft_chartqa_100step/checkpoints/99/items
================================================================================

------------------------------------------------------------
Checkpoint Freezing & Training Verification:
------------------------------------------------------------
  [PASS] Vision Encoder : 437/437 tensors frozen
  [PASS] Projector MLP  : 6/6 tensors updated (max diff: 0.010238)
  [PASS] LLM Backbone   : 13/13 tensors frozen
------------------------------------------------------------
```

---

# Quick Evaluation

Quick sample generation check with `decode_omni`:
```bash
export PYTHONPATH=src:$PYTHONPATH

python3 -m maxtext.experimental.omni_poc.utils.decode_omni \
  --config_path="src/maxtext/experimental/omni_poc/maxtext-omni-gemma3-qwen3.yml" \
  --checkpoint_path="gs://yuchenhou-maxtext-logs/omni-gemma3-qwen3/multimodal/sft_chartqa/omni_sft_chartqa_100step/checkpoints/99/items" \
  --num_samples=2 \
  --max_new_tokens=128
```

### Evaluation Output:
After 100 steps of training
```text
Sample 1 (Index 1309):
  Question: What was the average annual total income per household of those in the top decile group?
  Ground Truth: ['186600']
  Model Response: 125.66

Sample 2 (Index 228):
  Question: Which gender is represented by the red color line?
  Ground Truth: ['Male']
  Model Response: Female
```

---

# Step-by-Step Objectives
- [PR 4485](https://github.com/google/maxtext/pull/4485) Multi-Directory Checkpoint Restoration: Implemented selective sub-tree parameter loading using Orbax such that we can initialize a multimodal model with parameters from multiple checkpoints.
- [PR 4593](https://github.com/google/maxtext/pull/4593) Dynamic MLP Connector: Add omni modal adapter layer (MLP) to connect vision tower output to the LLM decoder.
- [PR 4746](https://github.com/google/maxtext/pull/4746) and [PR 4839](https://github.com/google/maxtext/pull/4839) Special Tokenizer & Placeholder Masking: Add special token `<|image_pad|>` to tokenizer and ensure correct masking in the decoder.
- [PR 4864](https://github.com/AI-Hypercomputer/maxtext/pull/4864) COCO-Narratives Data Pipeline Processing.
- **[Current] SFT Training & Checkpoint Verification**.
- SFT Evaluation & Benchmarking.

---

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/experimental/omni_poc/utils/decode_omni.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  true_length = len(combined_tokens)
+  seq_len = config.max_target_length
+  if true_length > seq_len:
+    raise ValueError(
+        f"The combined length of expanded prompt and vision tokens ({true_length}) "
+        f"exceeds config.max_target_length ({seq_len}). "
+        "Please increase max_target_length in your model config."
+    )
+  pad_token = tokenizer.pad_token_id if tokenizer.pad_token_id is not None else tokenizer.eos_token_id
+  padded_tokens = combined_tokens + [pad_token] * (seq_len - true_length)
```
[the remark this answers is not in the document]
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/decode_omni.py]: I think if both eos_token_id and pad_token are None in the tokenizer, then something must be wrong in tokenizer set up. It's better to make it trigger an error instead of silently failing

File: src/maxtext/experimental/omni_poc/tests/compare_sft_checkpoint_test.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+def _format_path(key_path) -> str:
+  parts = []
+  for k in key_path:
+    if hasattr(k, "key"):
+      parts.append(str(k.key))
+    elif hasattr(k, "idx"):
+      parts.append(str(k.idx))
+    else:
+      parts.append(str(k))
+  return "/".join(parts)
```
[the remark this answers is not in the document]
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/tests/compare_sft_checkpoint_test.py:89]: Sounds reasonable - updated

File: src/maxtext/experimental/omni_poc/tests/compare_sft_checkpoint_test.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+  projector_diffs = []
+  unexpected_diffs = []
+
+  # Go through all parameter tensors and compare weights before vs after SFT
+  for path, sft_arr in sft_leaves.items():
+    stitched_arr = stitched_leaves.get(path)
+    if stitched_arr is None:
+      print(f"Warning: Parameter path '{path}' not found in stitched checkpoint.")
+      continue
+
```
[the remark this answers is not in the document]
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/tests/compare_sft_checkpoint_test.py:199]: Agree with this - updated

File: src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+gradient_clipping_threshold: 1.0
+
+# -------------- HF pipeline (ibm-granite/ChartNet - human_verified) --------------
+# Dataset features in human_verified subset: ['id', 'image', 'code', 'csv', 'summary']
+# Prompt: 'prompt' (injected via default_prompt)
+# Completion: 'summary' (dense natural-language description of chart trends, axes, and data)
+dataset_type: hf
+tokenizer_type: "huggingface"
+tokenizer_path: "Qwen/Qwen3-4B"
+hf_path: 'ibm-granite/ChartNet'
```
Comment (hengtaoguo) [inline src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml:50]: Would you like to add a general ChartNet SFT config as a follow-up?

**Review by hengtaoguo — APPROVED** [review]
Thanks for the contribution! Approve for experimental merge

File: src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml (hunk 4 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/configs/pretrain-maxtext-omni-gemma3-qwen3-chartnet.yml:50]: Good idea - will do that as a separate PR! (update: [PR 5035](https://github.com/AI-Hypercomputer/maxtext/pull/5035))

**Review by NuojCheng — APPROVED** [review]
